File: plugins/context-tree/tests/visual_settings_ui.js
const assert = require("node:assert/strict");
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawn, spawnSync } = require("node:child_process");
const { chromium } = require("playwright");

const root = path.resolve(__dirname, "..");
const script = path.join(root, "scripts", "context_tree.py");
const artifacts = path.join(__dirname, "artifacts");
fs.mkdirSync(artifacts, { recursive: true });
const temporary = fs.mkdtempSync(path.join(os.tmpdir(), "context-tree-settings-"));
const store = path.join(temporary, ".context-tree");
const python = process.env.PYTHON || "python";
const server = spawn(python, [script, "--store", store, "ui", "--port", "0", "--no-open"], {
  stdio: ["ignore", "pipe", "pipe"],
  env: { ...process.env, PYTHONUTF8: "1", PYTHONIOENCODING: "utf-8" },
});

function firstLine(stream) {
  return new Promise((resolve, reject) => {
    let buffer = "";
    stream.setEncoding("utf8");
    stream.on("data", (chunk) => {
      buffer += chunk;
      const newline = buffer.indexOf("\n");
      if (newline >= 0) resolve(buffer.slice(0, newline).trim());
    });
    server.once("exit", (code) => reject(new Error(`settings server exited with ${code}`)));
  });
}

function stopServer() {
  if (server.exitCode === null) {
    if (process.platform === "win32") {
      spawnSync("taskkill", ["/pid", String(server.pid), "/t", "/f"], { stdio: "ignore" });
    } else {
      server.kill();
    }
  }
  server.stdout.destroy();
  server.stderr.destroy();
  server.unref();
}

(async () => {
  const url = await firstLine(server.stdout);
  const browser = await chromium.launch({ headless: true });
  try {
    const desktop = await browser.newPage({ viewport: { width: 1440, height: 900 } });
    await desktop.goto(url);
    const toggle = desktop.locator("#float-persistent");
    await toggle.waitFor();
    assert.equal(await toggle.isChecked(), false);
    assert.equal(await desktop.locator("#float-persistent-state").innerText(), "空闲时自动隐藏");
    await toggle.check();
    await desktop.waitForFunction(() => (
      document.querySelector("#float-persistent-state")?.textContent === "始终保留折叠按钮"
    ));
    assert.equal(await desktop.locator("#float-persistent-state").innerText(), "始终保留折叠按钮");
    assert.equal(await desktop.evaluate(() => document.documentElement.scrollWidth <= window.innerWidth), true);
    await desktop.screenshot({ path: path.join(artifacts, "settings-persistent-desktop.png"), fullPage: true });

    const mobile = await browser.newPage({ viewport: { width: 390, height: 844 } });
    await mobile.goto(url);
    await mobile.locator("#float-persistent").waitFor();
    await mobile.waitForFunction(() => document.querySelector("#float-persistent")?.checked === true);
    assert.equal(await mobile.locator("#float-persistent").isChecked(), true);
    assert.equal(await mobile.evaluate(() => document.documentElement.scrollWidth <= window.innerWidth), true);
    await mobile.screenshot({ path: path.join(artifacts, "settings-persistent-mobile.png"), fullPage: true });

    const widget = await browser.newPage({ viewport: { width: 380, height: 500 } });
    await widget.setContent(fs.readFileSync(path.join(root, "assets", "widget.html"), "utf8"));
    await widget.evaluate((value) => window.dispatchEvent(new MessageEvent("message", {
      source: window,
      data: { jsonrpc: "2.0", method: "ui/notifications/tool-result", params: { structuredContent: value } },
    })), {
      config: { consolidate_every: 5 }, topics: [], summary: { pending_count: 4 },
      usage: { mode: "real", capacity_tokens: 272000, sessions: [] },
      active_sessions: [{
        session_id: "thread-a", usage_session_id: "thread-a:7", session_title: "研究对话上下文延续",
        window_slot: "7", context_tokens: 202905, capacity_tokens: 353400, percent: 57.4,
        first_token_ms: 3823, compaction_count: 7, level: "warning",
        cost_profile_status: "partial_history", turn_cost_request_count: 105,
        turn_avg_cost: 0.07874957, turn_avg_context_cost: 0.07092929,
        turn_avg_first_token_ms: 4862.6, turn_avg_duration_ms: 10335,
        current_epoch_index: 10, current_epoch_avg_cost: 0.07874957,
        reference_epoch_label: "第3次压缩后", reference_avg_cost: 0.10919274,
        reference_epoch_extra_avg_cost: -0.03044317,
        attention_text: "历史较长，已使用增量读取；建议检查交接质量",
      }],
    });
    await widget.locator("#trigger").click();
    await widget.locator(".attention").waitFor();
    assert.equal(await widget.locator(".attention").innerText(), "历史较长，已使用增量读取；建议检查交接质量");
    assert.match(await widget.locator(".metrics").innerText(), /202,905 \/ 353,400/);
    assert.match(await widget.locator(".cost-profile").innerText(), /\$0\.0787/);
    assert.match(await widget.locator(".cost-profile").innerText(), /\$0\.1092/);
    assert.equal(await widget.evaluate(() => document.documentElement.scrollWidth <= window.innerWidth), true);
    await widget.screenshot({ path: path.join(artifacts, "widget-new-session-warning.png"), fullPage: true });
    console.log("Settings visual checks passed: desktop and mobile");
  } finally {
    await browser.close();
    stopServer();
    fs.rmSync(temporary, { recursive: true, force: true, maxRetries: 5, retryDelay: 100 });
  }
})().catch((error) => {
  stopServer();
  fs.rmSync(temporary, { recursive: true, force: true, maxRetries: 5, retryDelay: 100 });
  console.error(error);
  process.exitCode = 1;
});

File: plugins/context-tree/tests/visual_settings_ui.py
import os
from pathlib import Path

from playwright.sync_api import sync_playwright


ROOT = Path(__file__).resolve().parents[1]
ARTIFACTS = ROOT / "tests" / "artifacts"
ARTIFACTS.mkdir(parents=True, exist_ok=True)
URL = os.environ.get("CONTEXT_TREE_UI_URL", "http://127.0.0.1:18765/")
CHROME = os.environ.get(
    "CONTEXT_TREE_CHROME",
    r"C:\Program Files\Google\Chrome\Application\chrome.exe",
)


def assert_no_overlap(page) -> None:
    header = page.locator(".app-header").bounding_box()
    shell = page.locator(".shell").bounding_box()
    assert header and shell
    assert header["y"] + header["height"] <= shell["y"] + 1
    rows = page.locator(".topic-row").evaluate_all(
        """elements => elements.map(row => {
            const box = row.getBoundingClientRect();
            const button = row.querySelector('.select-topic').getBoundingClientRect();
            return {box: {x: box.x, width: box.width, height: box.height},
                    button: {x: button.x, width: button.width, height: button.height}};
        })"""
    )
    for row in rows:
        assert row["box"]["height"] >= 70
        assert row["button"]["height"] > 0
        assert row["button"]["x"] + row["button"]["width"] <= row["box"]["x"] + row["box"]["width"] + 1


with sync_playwright() as playwright:
    browser = playwright.chromium.launch(headless=True, executable_path=CHROME)
    page = browser.new_page(viewport={"width": 1440, "height": 900})
    page.goto(URL)
    page.wait_for_load_state("networkidle")
    assert page.locator("#usage-server").is_visible()
    assert page.locator("#backfill-session").is_visible()
    page.get_by_role("button", name="新建主题").click()
    page.locator("#topic-title").fill("本地主题切换")
    page.locator("#topic-summary").fill("不经过 AI 修改主题和整理频率")
    page.locator("#topic-form").get_by_role("button", name="创建").click()
    topic_row = page.locator(".topic-row", has_text="本地主题切换").first
    topic_row.wait_for()
    topic_row.locator(".select-topic").click()
    page.wait_for_function("document.querySelector('#selection-name').textContent === '本地主题切换'")
    assert page.locator("#selection-name").inner_text() == "本地主题切换"
    page.locator("#mode-auto").click()
    page.wait_for_function("document.querySelector('#selection-name').textContent === '自动判断'")
    page.locator("#intervals button[data-value='5']").click()
    page.wait_for_function("document.querySelector('#intervals button[data-value=\"5\"]').classList.contains('active')")
    assert page.locator("#selection-name").inner_text() == "自动判断"
    assert "active" in (page.locator("#intervals button[data-value='5']").get_attribute("class") or "")
    assert_no_overlap(page)
    page.screenshot(path=str(ARTIFACTS / "settings-backfill-desktop.png"), full_page=True)

    mobile = browser.new_page(viewport={"width": 390, "height": 844})
    mobile.goto(URL)
    mobile.wait_for_load_state("networkidle")
    assert mobile.locator("#usage-server").is_visible()
    assert mobile.locator("#backfill-session").is_visible()
    mobile.locator(".topic-title strong", has_text="本地主题切换").first.wait_for()
    assert_no_overlap(mobile)
    mobile.screenshot(path=str(ARTIFACTS / "settings-backfill-mobile.png"), full_page=True)
    browser.close()
